pandas/tests/test_categorical.py

```diff
@@ -94,13 +94,35 @@ def test_constructor_unsortable(self):
 
         # it works!
         arr = np.array([1, 2, 3, datetime.now()], dtype='O')
-        factor = Categorical.from_array(arr, ordered=False)
-        self.assertFalse(factor.ordered)
+        msg = "Categoricals cannot be object dtype unless all values are " \
+              "strings or all are periods."
+        with tm.assertRaisesRegexp(TypeError, msg):
+            factor = Categorical.from_array(arr, ordered=False)
 
         # this however will raise as cannot be sorted
         self.assertRaises(
             TypeError, lambda: Categorical.from_array(arr, ordered=True))
 
+    def test_constructor_object_dtype(self):
```

Comment:
can u add tests:

- unicode (for py2, include 2 bytes)
- str and unicode mixed (for py2, it's often the case in asian countries)
- bool (not sure it is used as category)
- period with mixed freq

I think some of them can't be covered because of  `infer_dytpe` spec and current impl.